lib/ptracer.h

```diff
@@ -0,0 +1,485 @@
+
+/* ---------------- DATA STRUCTURES ---------------- */
+
+/* What type of frame we are dealing with? Is it just a normal
+   C function or Lua C Function? */
+typedef enum frame_type {
+    PALLENE_TRACER_FRAME_TYPE_C,
+    PALLENE_TRACER_FRAME_TYPE_LUA
+} frame_type_t;
+
+/* Details of the callee function (name, where it is from etc.) */
+/* Optimization Tip: Try making it 'static' for every function. */
+typedef struct pt_fn_details {
+    const char *const fn_name;
+    const char *const filename;
+} pt_fn_details_t;
+
+/* A single frame representation. */
+typedef struct pt_frame {
+    frame_type_t type;
+    int line;
+
+    union {
+        pt_fn_details_t *details;
+        lua_CFunction frame_sig;
+    } shared;
+} pt_frame_t;
+
+/* Out stack is fully heap-allocated stack. We need some structure to hold
+   the stack information. This structure will be an Userdatum. */
+/* 'cont' stands for 'container'. */
+typedef struct pt_cont {
+    pt_frame_t *stack;
+    int count;
+} pt_cont_t;
+
+/* ---------------- DATA STRUCTURES END ---------------- */
+
+/* ---------------- DECLARATIONS ---------------- */
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+
+/* Initializes the Pallene Tracer. The initialization refers to creating the stack
+   if not created, preparing the traceback fn and finalizers. */
+/* NOTE: Pushes the finalizer object to the stack. The object has to be closed
+   everytime you are in a Lua C function using `lua_toclose(L, idx)` function. */
+PT_API pt_cont_t *pallene_tracer_init(lua_State *L);
+
+/* Pushes a frame to the stack. The frame structure is self-managed for every function. */
+PT_API void pallene_tracer_frameenter(lua_State *L, pt_cont_t *cont, pt_frame_t *restrict frame);
+
+/* Sets line number to the topmost frame in the stack. */
+PT_API void pallene_tracer_setline(pt_cont_t *cont, int line);
+
+/* Removes the last frame from the stack. */
+PT_API void pallene_tracer_frameexit(pt_cont_t *cont);
+
+/* Pallene Tracer explicit traceback function to show Pallene call-stack
+   tracebacks. */
+PT_API int  pallene_tracer_debug_traceback(lua_State *L);
+
+/* When we encounter an error, `pallene_tracer_frameexit()` may not
+   get called. Therefore, the stack will get corrupted if the previous
+   call-frames are not removed. The finalizer function makes sure that
+   it does not happen. Its guardian angel. */
+/* The finalizer function will be called from a to-be-closed value (since
+   Lua 5.4). If you are using Lua version prior 5.4, you are outta luck. */
+PT_API int pallene_tracer_finalizer(lua_State *L);
+
+/* When Pallene call-stack is overflowed. */
+l_noret pallene_tracer_runtime_callstack_overflow_error(lua_State *L);
+
+#ifdef __cplusplus
+}
+#endif // __cplusplus
+
+/* ---------------- DECLARATIONS END ---------------- */
+
+#endif // PALLENE_TRACER_H
+
+#if defined(PT_IMPLEMENTATION) && !defined(PT_IMPLEMENTED)
+#define PT_IMPLEMENTED
```

Comment:
This macro is not used anywhere.
I think what we want is something like this, with a single macro:

```cpp
#ifndef PT_IMPLEMENTATION
#define PT_IMPLEMENTATION

// implementation code goes here

#endif
```

Reply:
This macro is not meant to be used anywhere. It is used as redefinition protection guard for static implementation. For the record, if we would want to statically link `pallene-tracer` (that is if we want all the function definition to be included within the module instead of relying on some other shared library), we would do something like this: 

```C
#define PT_IMPLEMENTATION
#include <ptracer.h>
```

The implementation guard has to be a different macro (as we need to introduce a new state here) so that we don't get screwed when we do something like this: 
```C
#define PT_IMPLEMENTATION
#include <ptracer.h>

// Bunch of other module code...

// Ooops...
#define PT_IMPLEMENTATION
#include <ptracer.h>
```

We would get a a redefinition error when we would do something like this, but because `PT_IMPLEMENTED` would get defined the first time we inlcude ptracer.h with implementation, we don't have to worry about redefinition error.

Reply:
Ah, it's for the users, not us.
Even still, why not just:

```c
#ifndef PT_IMPLEMENTATION

// implementation code goes here

#endif
```

Reply:
First let's define user = The dev. developing a Lua C module.

Yes, that is the general way to define STB headers. But, as we would document static linking (I don't know, can I call it static linking interchangeably?), people/users/devs may mistakenly include ptracer.h with `PT_IMPLEMENTATION` twice. Consider this, I am working on a module with multiple C file with tracebacks:

a_module_file.c
```C
#define PT_IMPLEMENTATION
#include <ptracer.h>

// Code...
```
Good, no problem here.

another_module_file.c
```C
// We shouldn't/can't do that because the implementation has already been included.
#define PT_IMPLEMENTATION
#include <ptracer.h>

// Code ...
```

Now, the poor dev. would get redefinition error without having a clue of where did the error happen and why. The correct code should have been: 

another_module_file.c
```C
// All OK.
#include <ptracer.h>

// Code ...
```

The `PT_IMPLEMENTED` macro is not necessary per se, but we need it so that devs. can work with our header with implementation painlessly. The header guard macros are not used throughout the code as well, we don't get rid of them now do we? Moreover, it's just 1 line and couple of chars. extra, shouldn't be a problem.

> Ah, it's for the users, not us.

It's a matter of perspective. Pallene Tracer is for the devs. so that they have a good time debugging. If we consider them users, the entire project is for the users. So should we exclude everything then? :)

Reply:
I didn't think anyone would do use the same #define twice before including a single header lib.
But yeah I see your point. We can keep it like this